### ISSUE TITLE
cluster/aws/util.sh: kube-down doesn't respect custom VPC_ID

### ISSUE THREAD

**Issue by peterbourgon** [kubernetes/kubernetes#17219]
1. export custom VPC_ID 
2. Invoke kube-up on AWS
3. Verify cluster works
4. Invoke kube-down

Expected behavior: all created k8s resources are removed.

Actual behavior: the script exits almost immediately, having not deleted anything.

It appears that [kube-down blindly invokes get_vpc_id](https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/util.sh#L1152) but that should only happen if VPC_ID isn't already set. [kube-up has the correct behavior](https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/util.sh#L700-L702).

**Comment by peterbourgon** [comment]
This patch seems to resolve it, modulo style concerns.

```patch
diff --git a/cluster/aws/util.sh b/cluster/aws/util.sh
index 0c3f3a3..115abb0 100755
--- a/cluster/aws/util.sh
+++ b/cluster/aws/util.sh
@@ -1149,7 +1149,10 @@ function check-cluster() {
 }
 
 function kube-down {
-  local vpc_id=$(get_vpc_id)
+  local vpc_id=${VPC_ID}
+  if [[ "${vpc_id}" == "" ]]; then
+    vpc_id=$(get_vpc_id)
+  fi
   if [[ -n "${vpc_id}" ]]; then
     local elb_ids=$(get_elbs_in_vpc ${vpc_id})
     if [[ -n "${elb_ids}" ]]; then
```

**Comment by peterbourgon** [comment]
...but with that, it still tries to delete the VPC, which it shouldn't do...

**Comment by peterbourgon** [comment]
Additionally, when I create a k8s cluster with a declared VPC_ID, I can't create LoadBalancers (ELBs) with my services anymore:

```
. . . CreatingLoadBalancerFailed      Error creating load balancer (will retry): Failed to create load balancer for service default/helloworld: ValidationError: Either AvailabilityZones or SubnetIds must be specified
. . . CreatingLoadBalancerFailed      Error creating load balancer (will retry): Failed to create load balancer for service default/helloworld: Throttling: Rate exceeded
. . . CreatingLoadBalancerFailed      Error creating load balancer (will retry): Failed to create load balancer for service default/helloworld: Throttling: Rate exceeded
```

**Comment by davidopp** [comment]
@justinsb any thoughts on this?

**Comment by royklopper** [comment]
@peterbourgon Kubernetes checks the Amazon API for resources with the tag named "KubernetesCluster". This must resemble the name of the cluster. Hope this will solve your problem.

**Comment by peterbourgon** [comment]
@royklopper That's fine, but it doesn't address the original issue.

**Comment by yissachar** [comment]
I was able to get a little further by commenting out the lines that delete the VPC itself. Unfortunately, the script deletes all ELBs in the VPC regardless of whether they were created by Kubernetes or not. I ended up in a weird state with some resources deleted that I didn't want deleted, and some Kubernetes resources left around which I had to manually clean up.

**Comment by thenayr** [comment]
This seems slightly concerning at the very least.  I'm sure there are more than a few people who use custom VPC's (myself included) that would like to be able to tear down a cluster without clobbering any existing infrastructure inside of that VPC.  

I just experienced this same problem, brought up a second cluster in a custom VPC, proceeded to run ```kube-down``` to destroy the cluster and ended up completely clobbering an existing cluster in a different VPC.   Thought I missed some env vars, but double checked and they were all present.

**Comment by davidopp** [comment]
@justinsb Can you give an update on this? It is marked with 1.3 milestone.

**Comment by activars** [comment]
Seems to be a critical issue... If it's not warned ahead, it's likely to cause pain for the user.

**Comment by erictune** [comment]
@davidopp Is this a release blocker?

**Comment by erictune** [comment]
key question is whether this worked in 1.2 and stopped working, versus it is just super annoying but never worked.

**Comment by activars** [comment]
I suspect the execution in certain scenario could be dangerous:
- what if people want to leave the VPC untouched, but only destroy resources created by the kube-up.
- what if we want keep the network settings

After a few days investigation, I believe a decent kube-down never worked... (Including security group clean up, dhcp clean up, routing table clean up, master deletion when VPC id is specified)

Since I have stuff to ship, the short term solution I had was using Ansible wrapped around current kube-up, and destroy resources correctly. 

I can share the ansible if this helps - let me know.

**Comment by davidopp** [comment]
I discussed with @justinsb just now.

As indicated in the discussion thread here, this never worked. However, it's a pretty bad problem.

@justinsb said he would look into it tonight. We discussed three possible approaches for the 1.3 timeframe:
* actually fix it (the long-term solution)
* partially fix it, so we don't over-kill, but possibly leak resources
* total hack: just add something to kube-down.sh that, if you're running on AWS, prints a warning like "if you are using custom VPC_ID, this may <description of what badness may happen>", press "y" to continue.

**Comment by activars** [comment]
Can we make sure that it only destroys resources being labelled/tagged as KubernetesCluster with specified prefix? 

There's an edge case: before the creation if a tagged resources already exist before creation, it should not allow creating another cluster with the same tag.

AWS Role name seems to be hard coded value. So it probably should be left untouched or refactored to be named with the cluster prefix.

**Comment by erictune** [comment]
@activars 

It there any way you could send a PR?

**Comment by activars** [comment]
@erictune Yes, definitely.

I do have most of the logics in the hacking Ansible script - just a matter of translate into the bash - I've been creating/destroying clusters in a pre-configured VPC 10 times a day without having issues, it's probably straight forward to reproduce.

e.g. PR #27278 is a small change relevant to the topic (i.e. a preconfigured VPC would have a DHCP set already, we shouldn't change the settings).

Do you mind give me a week or two, since I have a bit priority at work, but I can sort this out at free time.
By looking the 1.3 milestone due date 24th June, I will try to make a PR before that date.

**Comment by justinsb** [comment]
Sorry for delay.  But I've reviewed this in detail now, and I no longer see it as a particularly critical problem.

- In normal operation, kube-up owns the VPC, and will delete resources correctly.
- We have an experimental and undocumented option to reuse a VPC, by setting `VPC_ID`.  In that case we will not tag the VPC, and kube-down won't do _anything_.  That's not great, but it's not dangerous (other than to your EC2 bills).
- If you change the kube-down script, you can get it to delete other resources.
- If you manually tag the VPC with the KubernetesCluster tag (indicating it is k8s owned), then kube-down will try to tear it down, including resources in that VPC whether or not tagged with the KubernetesCluster annotation.  Again, not great, but arguably not incorrect.
- If you give two clusters the same id (using KUBE_AWS_INSTANCE_PREFIX) in the same region, then kube-down will likely try to delete both  This applies even if you specify a custom VPC_ID for one of them.  If you don't use the VPC_ID option, it isn't possible to create two clusters with the same id.  @thenayr I believe this is what you hit.

I'm inclined to say that the best fix here is for kube-down to print "we didn't find a VPC for that cluster.  If you want to remove a shared cluster, please use X".  Given kube-up & kube-down are now in [maintenance mode](https://github.com/kubernetes/kubernetes/tree/master/cluster#deprecation-notice-this-directory-has-entered-maintainence-mode-and-will-not-be-accepting-new-providers-please-submit-new-automation-deployments-to-kube-deploy-deployments-in-this-directory-will-continue-to-be-maintained-and-supported-at-their-current-level-of-support) I think we should point people to just run `upup delete cluster --name <clusterid> --region <myregion>`, but kube-deploy is a "big tent" and we can have many options.

Downgrading to P2, on the basis that @thenayr 's problem caused the upgrade to P0, but I now hypothesize this to be caused by using the same cluster-id in conjunction with VPC_ID.  @thenayr do you think this was the problem?  If not do you have any idea what could have happened in your scenario?

I don't think this is a great scenario, but I think there are two failure scenarios:

1. we don't delete anything (and we should point uses to try a more capable tool)
2. if users specify VPC_ID it is possible to create two clusters with the same name, which will cause problems on kube-down (and I'd say that we've always planned to address these use cases in a supported way with kube-up v2).

**Comment by justinsb** [comment]
And I opened #27518, which warns in the first scenario, and links to some task-based docs I am proposing adding to the kube-deploy repo (https://github.com/kubernetes/kube-deploy/pull/114)

**Comment by thenayr** [comment]
@justinsb Thanks for the lengthy update.

I do believe you are correct in your assumption that the `KUBE_AWS_INSTANCE_PREFIX` was the same across both clusters.  I didn't realize at the time the extent to which the instance prefix is referenced throughout the scripts.  

I still think the default behavior if a user specifies a custom `VPC_ID` to the kube-down script should be to retain the VPC and not attempt to tear it down.

**Comment by erictune** [comment]
Thanks for the update, Justin.

**Comment by justinsb** [comment]
> I still think the default behavior if a user specifies a custom VPC_ID to the kube-down script should be to retain the VPC and not attempt to tear it down.

I believe that is the behaviour.   kube-down ignores VPC_ID, doesn't find a tagged VPC, and doesn't destroy anything.  With #27518 it prints a hint as to why it isn't doing anything and points users to the more capable tools being developed in kube-deploy.

**Comment by goltermann** [comment]
Moving P2 code bugs out of v1.3 milestone.  Please comment if you disagree and think a fix is required for 1.3.

**Comment by justinsb** [comment]
We're not going to be making further changes to kube-up.  The behaviour of kube-up is now to fail-safe, I believe.  kops can delete a cluster safely (even with a shared VPC), including the case where it did not create a cluster (using the `--external` flag).  Not entirely sure whether there's something I'm missing that remains to-do on this issue, but I'm sure someone will let me know if so ;-)  Closing
